### PR TITLE
Update slots to reference before_render

### PR DESCRIPTION
**Pull request by brittanyellich**
### What are you trying to accomplish?

I recently ran into an issue using slots in view components that I think could save other people some time in the future by adding how to solve it to the documentation. Shoutout to @joelhawksley for showing me how to properly accomplish this!

### What approach did you choose and why?

I added an example that I think follows the theme of the rest of the doc to describe how to use before_render.

### Anything you want to highlight for special attention from reviewers?

Can I use class_names in the example here? And is it a good idea to reference the lifecycle.md page as well?

### CODE REVIEW

File: docs/guide/slots.md [hunk 1]
```diff
@@ -145,6 +145,51 @@ end
...
+  renders_one :header
+  renders_one :image
+  renders_many :posts
+
+  def initialize(**system_arguments)
+    @system_arguments = system_arguments
+
+    @post_container_classes = class_names(
+      "PostContainer",
+      system_arguments[:classes]
```
Comment (Spone) [inline docs/guide/slots.md:164]: Can we simplify a bit by removing `system_arguments`?


```suggestion
  def initialize(classes: [])
    @post_container_classes = class_names("post-container", classes)
```

What do you think?

File: docs/guide/slots.md [hunk 2]
```diff
@@ -145,6 +145,51 @@ end
 <% end %>
 ```
 
+## Referencing slots
+
+You can access the instance of a slot in your ruby file by using the `before_render` [lifecycle method](./lifecycle.md).
```
Comment (Spone) [inline docs/guide/slots.md:150]:
```suggestion
You can access the instance of a slot in your component class by using the `before_render` [lifecycle method](./lifecycle.md).
```

File: docs/guide/slots.md (hunk 1 above)
Comment (brittanyellich) [reply docs/guide/slots.md:164]: Absolutely!

File: docs/guide/slots.md (hunk 2 above)
Comment (brittanyellich) [reply docs/guide/slots.md:150]: Good suggestion, I always trip over this type of terminology :)

File: docs/guide/slots.md [hunk 3]
```diff
@@ -145,6 +145,47 @@ end
 <% end %>
 ```
 
+## Referencing slots
+
+You can access the instance of a slot in your component class by using the `before_render` [lifecycle method](./lifecycle.md).
```
Comment (joelhawksley) [inline docs/guide/slots.md:150]:
```suggestion
As slots are registered after initialization, use the `before_render` [lifecycle method](./lifecycle.md) to reference them:
```

File: docs/guide/slots.md [hunk 4]
```diff
@@ -145,6 +145,47 @@ end
...
+      @post_container_classes
+    )
+  end
+end
+```
+
+```erb
+<%# blog_component.html.erb %>
+<h1><%= header %></h1>
+
```
Comment (joelhawksley) [inline docs/guide/slots.md:178]:
```suggestion
```

File: docs/guide/slots.md [hunk 5]
```diff
@@ -145,6 +145,47 @@ end
...
+<% posts.each do |post| %>
+  <div class="<%= @post_container_classes =%>">
```
Comment (joelhawksley) [inline docs/guide/slots.md:180]:
```suggestion
  <div class="<%= @post_container_classes %>">
```

File: docs/guide/slots.md [hunk 6]
```diff
@@ -145,6 +145,47 @@ end
...
 ```
 
+## Referencing slots
+
+You can access the instance of a slot in your component class by using the `before_render` [lifecycle method](./lifecycle.md).
+
+```ruby
+# blog_component.rb
+class BlogComponent < ViewComponent::Base
+  renders_one :header
```
Comment (joelhawksley) [inline docs/guide/slots.md:155]:
```suggestion
# blog_component.rb
class BlogComponent < ViewComponent::Base
```

File: docs/guide/slots.md [hunk 7]
```diff
@@ -145,6 +145,47 @@ end
...
+  renders_one :image
+  renders_many :posts
+
+  def initialize(classes: [])
+    @post_container_classes = class_names("post-container", classes)
+    )
+  end
+
```
Comment (joelhawksley) [inline docs/guide/slots.md:163]:
```suggestion
```

File: docs/guide/slots.md [hunk 8]
```diff
@@ -145,6 +145,47 @@ end
...
+  def before_render
+    @post_container_classes = class_names(
+      {
+        "PostContainer--hasImage": image.present?
+      },
+      @post_container_classes
+    )
```
Comment (joelhawksley) [inline docs/guide/slots.md:170]:
```suggestion
    "PostContainer--hasImage" if image.present?
```

**Review by joelhawksley — APPROVED** [review]
👋🏻 looks good! I think we can simplify this example code a bit. I'll take it from here ❤️

File: docs/guide/slots.md [hunk 9]
```diff
@@ -145,6 +145,34 @@ end
...
+end
+```
+
+```erb
+<%# blog_component.html.erb %>
+<% posts.each do |post| %>
+  <div class="<%= @post_container_classes %>">
+    <% if image? %>
+      <%= image %>
+    <% end %>
```
Comment (joelhawksley) [inline docs/guide/slots.md:170]:
```suggestion
    <%= image if image? %>
```

Comment (joelhawksley) [inline docs/guide/slots.md:170]:
```suggestion
    <%= image if image? %>
```

File: docs/guide/slots.md [hunk 10]
```diff
@@ -145,6 +145,32 @@ end
...
+As the content passed to slots is registered after a component is initialized, it referenced in an initializer. One way to reference slots is using the `before_render` [lifecycle method](./lifecycle.md):
+
+```ruby
+# blog_component.rb
+class BlogComponent < ViewComponent::Base
+  renders_one :image
+  renders_many :posts
+
+  def before_render
+    "PostContainer--hasImage" if image.present?
```
Comment (joelhawksley) [inline docs/guide/slots.md:159]:
```suggestion
    @post_container_classes = "PostContainer--hasImage" if image.present?
```

File: docs/guide/slots.md [hunk 11]
```diff
@@ -145,6 +145,32 @@ end
 <% end %>
 ```
 
+## Referencing slots
+
+As the content passed to slots is registered after a component is initialized, it referenced in an initializer. One way to reference slots is using the `before_render` [lifecycle method](./lifecycle.md):
```
Comment (joelhawksley) [inline docs/guide/slots.md:150]:
```suggestion
As the content passed to slots is registered after a component is initialized, it cannot referenced in an initializer. One way to reference slot content is using the `before_render` [lifecycle method](./lifecycle.md):
```

File: docs/guide/slots.md [hunk 12]
```diff
@@ -145,6 +145,32 @@ end
 <% end %>
 ```
 
+## Referencing slots
+
+As the content passed to slots is registered after a component is initialized, it cannot referenced in an initializer. One way to reference slot content is using the `before_render` [lifecycle method](./lifecycle.md):
```
Comment (joelhawksley) [inline docs/guide/slots.md:150]:
```suggestion
As the content passed to slots is registered after a component is initialized, it cannot referenced in an initializer. One way to reference slot content is using the `before_render` [lifecycle method](/guide/lifecycle):
```

File: docs/guide/slots.md [hunk 13]
```diff
@@ -145,6 +145,32 @@ end
 <% end %>
 ```
 
+## Referencing slots
+
+As the content passed to slots is registered after a component is initialized, it cannot referenced in an initializer. One way to reference slot content is using the `before_render` [lifecycle method](/guide/lifecycle):
```
Comment (Spone) [inline docs/guide/slots.md:150]:
```suggestion
As the content passed to slots is registered after a component is initialized, it cannot be referenced in an initializer. One way to reference slot content is using the `before_render` [lifecycle method](/guide/lifecycle):
```

**Pull request merged**